src/zarr/core/dtype/__init__.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, TypeAlias, get_args
+
+if TYPE_CHECKING:
+    from zarr.core.common import ZarrFormat
+
+import numpy as np
+import numpy.typing as npt
+
+from zarr.core.common import JSON
+from zarr.core.dtype._numpy import (
+    _NUMPY_SUPPORTS_VLEN_STRING,
+    Bool,
+    Complex64,
+    Complex128,
+    DateTime64,
+    FixedLengthAscii,
+    FixedLengthBytes,
+    FixedLengthUnicode,
+    Float16,
+    Float32,
+    Float64,
+    Int8,
+    Int16,
+    Int32,
+    Int64,
+    Structured,
+    UInt8,
+    UInt16,
+    UInt32,
+    UInt64,
+    VariableLengthString,
+)
+from zarr.core.dtype.registry import DataTypeRegistry
+from zarr.core.dtype.wrapper import ZDType, _BaseDType, _BaseScalar
+
+__all__ = [
+    "Complex64",
+    "Complex128",
+    "DateTime64",
+    "FixedLengthAscii",
+    "FixedLengthBytes",
+    "FixedLengthUnicode",
+    "Float16",
+    "Float32",
+    "Float64",
+    "Int8",
+    "Int16",
+    "Int32",
+    "Int64",
+    "Structured",
+    "UInt8",
+    "UInt16",
+    "UInt32",
+    "UInt64",
+    "VariableLengthString",
+    "ZDType",
+    "data_type_registry",
+    "parse_data_type",
+]
+
+data_type_registry = DataTypeRegistry()
+
+INTEGER_DTYPE = Int8 | Int16 | Int32 | Int64 | UInt8 | UInt16 | UInt32 | UInt64
+FLOAT_DTYPE = Float16 | Float32 | Float64
+COMPLEX_DTYPE = Complex64 | Complex128
+STRING_DTYPE = FixedLengthUnicode | VariableLengthString | FixedLengthAscii
+DTYPE = (
+    Bool
+    | INTEGER_DTYPE
+    | FLOAT_DTYPE
+    | COMPLEX_DTYPE
+    | STRING_DTYPE
+    | FixedLengthBytes
+    | Structured
+    | DateTime64
+)
+
+ZDTypeLike: TypeAlias = npt.DTypeLike | ZDType[_BaseDType, _BaseScalar] | dict[str, JSON]
+
+for dtype in get_args(DTYPE):
+    data_type_registry.register(dtype._zarr_v3_name, dtype)
+
+
+# TODO: find a better name for this function
+def get_data_type_from_native_dtype(dtype: npt.DTypeLike) -> ZDType[_BaseDType, _BaseScalar]:
```

Comment:
```suggestion
def zarr_dtype_from_numpy_dtype(dtype: npt.DTypeLike) -> ZDType[_BaseDType, _BaseScalar]:
```
?

Reply:
we don't want to over-fit to numpy data types.

Reply:
if we add support for non-numpy data types (like arrow data types) then we don't have to rename this function.